File: crates/blazen-py/examples/llm_mistralrs.py
"""Local LLM inference with mistral.rs (no API key needed).

mistral.rs runs LLM inference entirely on-device using optimised Rust kernels.
It supports HuggingFace model IDs (automatic download + caching) and local GGUF
files.  No API key, no network access (after initial model download), and no
usage-based billing.

The first run downloads model weights from HuggingFace Hub and caches them
locally.  Subsequent runs reuse the cached weights automatically.

Unlike the cloud-based providers (OpenAI, Anthropic, etc.), mistral.rs is
ideal for local development, CI pipelines, air-gapped deployments, and
on-device inference where data must not leave the machine.

Usage:
    uv run python crates/blazen-py/examples/llm_mistralrs.py
"""

import asyncio

from blazen import ChatMessage, CompletionModel, CompletionOptions, MistralRsOptions


async def main() -> None:
    # ------------------------------------------------------------------
    # 1. Create a local completion model with mistral.rs
    # ------------------------------------------------------------------
    # Uses a small GGUF-quantised model suitable for quick demos.
    # Replace with any HuggingFace model ID or local GGUF path.
    opts = MistralRsOptions("TheBloke/Mistral-7B-Instruct-v0.2-GGUF")
    model = CompletionModel.mistralrs(options=opts)

    print(f"Model ID: {model.model_id}")
    print()

    # ------------------------------------------------------------------
    # 2. Simple completion
    # ------------------------------------------------------------------
    print("--- Simple Completion ---")
    response = await model.complete([
        ChatMessage.system(
            "You are a helpful assistant. Answer concisely in 1-2 sentences."
        ),
        ChatMessage.user("What is 2+2?"),
    ])

    print(f"Response: {response.content}")
    print(f"Model:    {response.model}")
    if response.usage:
        print(f"Tokens:   {response.usage}")
    print()

    # ------------------------------------------------------------------
    # 3. Completion with options (temperature, max_tokens)
    # ------------------------------------------------------------------
    print("--- Completion With Options ---")
    response = await model.complete(
        [
            ChatMessage.system("You are a creative storyteller."),
            ChatMessage.user("Tell me a very short story about a robot learning to cook."),
        ],
        CompletionOptions(temperature=0.9, max_tokens=256),
    )

    print(f"Response: {response.content}")
    print()

    # ------------------------------------------------------------------
    # 4. Streaming completion
    # ------------------------------------------------------------------
    print("--- Streaming ---")

    chunks_received = 0

    async for chunk in model.stream([
        ChatMessage.user("Explain why the sky is blue in 2-3 sentences."),
    ]):
        if chunk.delta:
            print(chunk.delta, end="", flush=True)
        chunks_received += 1

    print()
    print(f"\n(received {chunks_received} chunks)")
    print()

    print("Done.")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/blazen-node/examples/llm_mistralrs.ts
/**
 * Local LLM inference with Blazen's mistral.rs backend.
 *
 * Demonstrates using Blazen's CompletionModel with the mistral.rs engine to
 * run LLM inference entirely on-device -- no API key or network access
 * required (after the initial model download from HuggingFace Hub).
 *
 * The default model used here is a GGUF-quantised Mistral-7B-Instruct variant.
 * Replace with any HuggingFace model ID or local GGUF path.
 *
 * Run with: npx tsx llm_mistralrs.ts
 */

import { CompletionModel, ChatMessage } from "blazen";

// ---------------------------------------------------------------------------
// 1. Create a local mistral.rs model.
// ---------------------------------------------------------------------------

const model: CompletionModel = CompletionModel.mistralrs({
  modelId: "TheBloke/Mistral-7B-Instruct-v0.2-GGUF",
});

console.log(`Model ID: ${model.modelId}`);
console.log();

// ---------------------------------------------------------------------------
// 2. Simple completion.
// ---------------------------------------------------------------------------

console.log("--- Simple Completion ---");

const response = await model.complete([
  ChatMessage.system("You are a helpful assistant. Answer concisely in 1-2 sentences."),
  ChatMessage.user("What is 2+2?"),
]);

console.log(`Response: ${response.content}`);
console.log(`Model:    ${response.model}`);
if (response.usage) {
  console.log(`Tokens:   ${JSON.stringify(response.usage)}`);
}
console.log();

// ---------------------------------------------------------------------------
// 3. Completion with options (temperature, maxTokens).
// ---------------------------------------------------------------------------

console.log("--- Completion With Options ---");

const storyResponse = await model.completeWithOptions(
  [
    ChatMessage.system("You are a creative storyteller."),
    ChatMessage.user("Tell me a very short story about a robot learning to cook."),
  ],
  { temperature: 0.9, maxTokens: 256 },
);

console.log(`Response: ${storyResponse.content}`);
console.log();

// ---------------------------------------------------------------------------
// 4. Streaming completion.
// ---------------------------------------------------------------------------

console.log("--- Streaming ---");

let chunksReceived = 0;
await model.stream(
  [ChatMessage.user("Explain why the sky is blue in 2-3 sentences.")],
  (chunk) => {
    if (chunk.delta) {
      process.stdout.write(chunk.delta);
    }
    chunksReceived++;
  },
);

console.log();
console.log(`\n(received ${chunksReceived} chunks)`);
console.log();
console.log("Done.");
